repository/src/lib.rs

```diff
@@ -156,13 +156,184 @@ impl<T: RawRepository> DistributedRepository<T> {
+            let ancestor_commits = self
+                .raw
+                .list_ancestors(branch_tip_commit_hash, Some(256))
+                .await?;
+            let position = ancestor_commits
+                .iter()
+                .position(|c| *c == finalized_branch_commit_hash)
+                .expect("TODO: handle the case where it exceeds the limit");
+            let new_semantic_commits = stream::iter(
+                ancestor_commits
+                    .iter()
+                    .take(position)
+                    .rev()
+                    .cloned()
+                    .map(|c| {
+                        let raw = &self.raw;
+                        async move { raw.read_semantic_commit(c).await.map(|x| (x, c)) }
+                    }),
+            )
+            .buffered(256)
+            .collect::<Vec<_>>()
+            .await;
+            let mut new_semantic_commits = new_semantic_commits
+                .into_iter()
+                .collect::<Result<Vec<_>, _>>()?;
+            // Add most recent commit of the branch to the list since it is not included in the ancestor commits
+            let branch_tip_semantic_commit = self
+                .raw
+                .read_semantic_commit(branch_tip_commit_hash)
+                .await?;
+            new_semantic_commits.push((branch_tip_semantic_commit, branch_tip_commit_hash));
+            let new_commits = new_semantic_commits
+                .into_iter()
+                .map(|(commit, hash)| {
+                    from_semantic_commit(commit)
+                        .map_err(|e| (e, hash))
+                        .map(|x| (x, hash))
+                })
+                .collect::<Result<Vec<_>, _>>()
+                .map_err(|(error, hash)| {
+                    anyhow!("failed to convert the commit {}: {}", hash, error)
+                })?;
+
+            // Verify all the incoming commits and apply them to the local repository only if they are valid.
+            let mut verifier =
+                CommitSequenceVerifier::new(last_header.clone(), reserved_state.clone())
+                    .map_err(|e| anyhow!("failed to create a verifier: {}", e))?;
+            for (new_commit, _) in &new_commits {
+                if verifier.apply_commit(new_commit).is_err() {
+                    continue 'branch;
+                }
+            }
+            let branch_tip_commit = &new_commits.last().unwrap().0;
+            if let Commit::Agenda(_) = branch_tip_commit {
+                // If branch tip commit is Agenda commit, create a new agenda branch
+                let branch_tip_commit_hash = self.raw.locate_branch(branch).await?;
+                let branch_name = "a-"; //TODO : change # into number
+                let branch_name = format!("{}{}", branch_name, branch_tip_commit_hash);
```

Comment:
the number after `a-` should be the truncated 8bits of of the hash value of agenda commit. not the branch tip commit hash.